src/gleam/time/duration.gleam

```diff
@@ -0,0 +1,175 @@
+import gleam/int
+import gleam/order
+import gleam/string
+
+/// An amount of time, with up to nanosecond precision.
+///
+/// This type does not represent calendar periods such as "1 month" or "2
+/// days". Those periods will be different lengths of time depending on which
+/// month or day they apply to. For example, January is longer than February.
+/// A different type should be used for calendar periods.
+///
+pub opaque type Duration {
+  // When compiling to JavaScript ints have limited precision and size. This
+  // means that if we were to store the the timestamp in a single int the
+  // duration would not be able to represent very large or small durations.
+  // Durations are instead represented as a number of seconds and a number of
+  // nanoseconds.
+  //
+  // If you have manually adjusted the seconds and nanoseconds values the
+  // `normalise` function can be used to ensure the time is represented the
+  // intended way, with `nanoseconds` being positive and less than 1 second.
+  Duration(seconds: Int, nanoseconds: Int)
+}
+
+/// Ensure the duration is represented with `nanoseconds` being positive and
+/// less than 1 second.
+///
+/// This function does not change the amount of time that the duratoin refers
+/// to, it only adjusts the values used to represent the time.
+///
+fn normalise(duration: Duration) -> Duration {
+  let multiplier = 1_000_000_000
+  let nanoseconds = duration.nanoseconds % multiplier
+  let overflow = duration.nanoseconds - nanoseconds
+  let seconds = duration.seconds + overflow / multiplier
+  Duration(seconds, nanoseconds)
+}
+
+/// Compare one duration to another, indicating whether the first is greater or
+/// smaller than the second.
+///
+/// # Examples
+///
+/// ```gleam
+/// compare(seconds(1), seconds(2))
+/// // -> order.Lt
+/// ```
+///
+pub fn compare(left: Duration, right: Duration) -> order.Order {
+  order.break_tie(
+    int.compare(left.seconds, right.seconds),
+    int.compare(left.nanoseconds, right.nanoseconds),
+  )
+}
```

Comment:
I think we could add more documentation here explaining how the comparison is working. If the "greater" means "X this took longer than Y" we should say it explicitly and show more examples, with a mix of negative and positive values
(and add tests for each!)